python/GafferTest/NodeAlgoTest.py

```diff
@@ -361,6 +361,89 @@ def testBadVisitorReturnValue( self ) :
 		with self.assertRaisesRegexp( RuntimeError, "Visitor must return a bool \(True to continue, False to prune\)" ) :
 			Gaffer.NodeAlgo.visitUpstream( g["L3_1"], lambda node : None )
 
+	def __boxedVisitationGraph( self ) :
+
+		s = self.__visitationGraph()
+		b = Gaffer.Box.create( s, Gaffer.StandardSet( [ s["L1_1"] ] ) )
+		b.setName( "Box_L1_1" )
+		b = Gaffer.Box.create( s, Gaffer.StandardSet( [ s["L2_3"] ] ) )
+		b.setName( "Box_L2_3" )
+		return s
+
+	def testVisitBoxedNodesDepthFirst( self ) :
+
+		s = self.__boxedVisitationGraph()
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["Box_L1_1"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s[ "L2_1"], s["L3_1"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["Box_L1_1"]["L1_1"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s["Box_L1_1"], s[ "L2_1"], s["L3_1"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["L1_2"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s["L2_2"], s["L3_2"], s["Box_L2_3"], s["Box_L2_3"]["L2_3"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["Box_L2_3"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s[ "L1_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["Box_L2_3"]["L2_3"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s["Box_L2_3"], s[ "L1_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.upstreamNodes( s["L3_1"], order = Gaffer.NodeAlgo.VisitOrder.DepthFirst ),
+			[ s[ "L2_1"], s["Box_L1_1"], s["Box_L1_1"]["L1_1"] ],
+		)
+
+	def testVisitBoxedNodesBreadthFirst( self ) :
+
+		s = self.__boxedVisitationGraph()
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["Box_L1_1"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s[ "L2_1"], s["L3_1"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["Box_L1_1"]["L1_1"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s["Box_L1_1"], s[ "L2_1"], s["L3_1"], s["L3_2"] ],
+		)
+
+		self.assertEqual(
+			Gaffer.NodeAlgo.downstreamNodes( s["L1_2"], order = Gaffer.NodeAlgo.VisitOrder.BreadthFirst ),
+			[ s["L2_2"], s["Box_L2_3"], s["Box_L2_3"]["L2_3"], s["L3_2"] ],
```

Comment:
Do we need a check that BFS/DFS inside a box with multiple nodes is occurring correctly? It would be good to document the expected order for boxes too. As we have it makes sense, but someone may be under the impression that a DFS had consistent path length for each 'pass', instead of that only being measured at the outer level?

Reply:
> Do we need a check that BFS/DFS inside a box with multiple nodes is occurring correctly?

I don't think it adds all that much value over the existing tests, but I've added one in 5a3275f.

> It would be good to document the expected order for boxes too.

I must confess, I deliberately hedged that one. The current ordering is what naturally fell out of the most obvious implementation, and happily turned out to be what I wanted for the EditScope menu, but I'm not sure it stands up to rigorous scrutiny. Rather than document a specific behaviour, I want to wait and see if any other use cases prove me wrong.

Reply:
Hehe, fair enough :) Thanks for adding the test too.